diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index 146adc66b..da27863de 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -1038,7 +1038,11 @@ struct rq {
 	 * one CPU and if it got migrated afterwards it may decrease
 	 * it on another CPU. Always updated under the runqueue lock:
 	 */
-	unsigned long 		nr_uninterruptible;
+#ifdef __GENKSYMS__
+	unsigned int		nr_uninterruptible;
+#else
+	unsigned long		nr_uninterruptible;
+#endif
 
 	struct task_struct __rcu	*curr;
 	struct task_struct	*idle;
